### PR TITLE
DAOS-13540 object: handle TX_RESTART during TX convert

### PR DESCRIPTION
**Pull request by Nasf-Fan**
During TX convert, if -DER_TX_RESTART is returned, in spite of it is caused by stale pool map or transaction conflict, the TX convert logic needs to restart the internal transaction instead of forwarding the failure to upper layer caller.

### Before requesting gatekeeper:

* [ ] Two review approvals and any prior change requests have been resolved.
* [ ] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [ ] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [ ] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [ ] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate watchers.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficient testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

### CODE REVIEW

**Comment by Nasf-Fan** [comment]
Ping reviewers, thanks!

**Review by wangdi1 — APPROVED** [review]

File: src/object/obj_tx.c [hunk 1]
```diff
@@ -1041,71 +1052,60 @@ dc_tx_commit_cb(tse_task_t *task, void *data)
...
-	tx->tx_set_resend = 1;
-	tcca->tcca_args->flags |= DTF_RETRY_COMMIT;
-
 	D_MUTEX_UNLOCK(&tx->tx_lock);
 	locked = false;
 
-	if (pool_task != NULL) {
-		rc1 = dc_task_depend(task, 1, &pool_task);
+	if (rc != -DER_TX_RESTART) {
+		rc1 = dc_task_resched(task);
```
Comment (liuxuezhao) [inline src/object/obj_tx.c:1086]: I don't quite understand the logic, here why for "rc != -DER_TX_RESTART" case need to reinit the task? (I thought for rc == -DER_TX_RESTART case should reinit it?)

File: src/object/obj_tx.c [hunk 2]
```diff
@@ -3441,14 +3446,14 @@ dc_tx_attach(daos_handle_t th, struct dc_object *obj, enum obj_rpc_opc opc,
 		if (pu->flags & DAOS_COND_PUNCH) {
 			tx->tx_has_cond = 1;
 			D_MUTEX_UNLOCK(&tx->tx_lock);
-			obj_decref(obj);
+			if (comp)
+				obj_decref(obj);
```
[the remark this answers is not in the document]
Comment (Nasf-Fan) [reply src/object/obj_tx.c:3450]: It is only for one case. The object reference also can be held via another path: dc_tx_convert_post() => dx_tx_convert_cb() => dc_tx_convert_restart() => dc_tx_attach().
I will add comment when have chance to refresh the patch or in subsequent patch.

File: src/object/obj_tx.c (hunk 1 above)
Comment (Nasf-Fan) [reply src/object/obj_tx.c:1086]: For DER_TX_RESTART restart case, the original task needs to re-execute related IO with new epoch and then commit. Here if we directly re-schedule the task, it only re-executes the commit task.

**Review by liuxuezhao — APPROVED** [review]

**Comment by Nasf-Fan** [comment]
Ping @daos-stack/daos-gatekeeper

**Pull request merged**